### DISCUSSION TITLE
parse until a string is found

### DISCUSSION THREAD

**Discussion by pth14** [louthy/language-ext#1117] [Q&A]
Hi,
I am still learning the lib.
I am trying to do something simple but I failed to do it so far: I am looking for a parser like noneof but for a string, not for a char or some chars.
In other words, I would like to skip all chars until a string is matched (this string should not be consumed, using attempt probably).

**Comment by louthy** [comment]
This suggests a poorly structured parser.  You really should be moving forwards if you are able to parse something expected.  Usually, you'd only backtrack when there is something unexpected in the stream and you need to try alternatives.

Perhaps if you expand your question to some more specifics, I can give you some better advice?

But, if you reaally want to do it, you can use `manyUntil`:

```c#
var p = manyUntil(anyChar, str("THING"));
var r = p.Parse("****THING").ToEither();          // [ '*', '*', '*', '*', '*' ]
```

Be warned though, it's a recursive operator - which means if the amount of text being skipped is too large, you'll probably get a stack overflow.  You can see [the implementation here](https://github.com/louthy/language-ext/blob/main/LanguageExt.Parsec/Parsers/Prim.cs#L790).

**Reply by pth14** [reply]
Sure, for example, I am parsing C source files.
I am interested in parsing enum and struct (in a first time).
I wrote parsers to parse it.
But when I read a .h or .c file, there are lot of lines before and after structs and enums that I want to skip, like some comments, #ifdef , ...
So far, my parsers don'twork if I apply it on the whole source files.

**Reply by pth14** [reply]
It doesn't work for me because I would like it to work with the following input:

```
var p = manyUntil(anyChar, str("THING"));
var r = p.Parse("****THATTHING").ToEither(); 
```

I get the following error: "error at (line 1, column 8): unexpected "A", expecting 'I'".

**Reply by pth14** [reply]
I really have a hard time composing with "many" and "either".
I try to parse code like the following (I added line numbers):

```
1. /* SystemInterface.h
2. *      SystemInterface
3. *       
4. */
5.
6. #ifndef SYSTEMTYPEDEFS_H_
7. #define SYSTEMTYPEDEFS_H_
8.
9. #include <stdint.h>
10.
11. /* Project Parameters
12.  * *********************************************************************************************************************************************************/
13.
14. enum { e_1_1 = 0 , e_1_2 = 1,}; typedef uint16_t e1;
15. enum { e_2_1 = 0, e_2_2 = 1,}; typedef uint16_t e2;
```

I wrote 1 parser to skip comments, 1 to skip pre processor statements and 1 to parse enums.
So far so good.
If I applied it sequentially (i.e. comments parser then pre processor parser then enums parser ) and in right order, it works.
The problem is I don't know the order and how many times it repeats(like in lines 11-12 where comments repeat).
It can repeat anywhere so I wrote a parser like this:

```

EnumsParser =
  from e in many(either(spacesP, either(multiLinesCommentP, either(preProcessorP, EnumParser))))
  select e.Where(er => er.IsValid).AsEnumerable();
```

parsers ending in "P" are just parsers that return all the same type to allow to use "either" (IsValid property allow me to filter, getting only enum data).
I get this error: "unexpected many: combinator 'many' is applied to a parser that accepts an empty string."

**Reply by louthy** [reply]
`spacesP` presumably allows *zero* or more spaces, that means it will parse forever if you put it in a `many` (because parsing zero spaces will always succeed) - that's why you get the error.  If you put something in a `many` it _must_ parse something or fail, it can't successfully parse nothing.  For example, using something like `many1(satisfy(Char.IsWhiteSpace))` for `spacesP`, would only succeed when it parses one or more space.

You can also use `choice` instead of `either` when you have more than two options:
```c#
    many(choice(spacesP, multiLinesCommentP, preProcessorP, EnumParser))
```

**Reply by louthy** [reply]
btw, if you're looking for something to reference.  [I wrote a C# parser a few years back with LanguageExt.Parsec](https://github.com/louthy/best-form/blob/f9f5f67d2d5a6d3ffa8d5bedb6371fda84f37648/bestform.cs/Lang.cs), for my documentation generator.  It also tries to extract comments (for the documentation).  Note, it will skip over the code blocks as it's mostly trying to extract type and member definitions (and their comments).  

I deprecated it and switched to using Roslyn (as C# kept changing!), so this is a link to an old commit has.

**Reply by pth14** [reply]
> `spacesP` presumably allows _zero_ or more spaces, that means it will parse forever if you put it in a `many` (because parsing zero spaces will always succeed) - that's why you get the error. If you put something in a `many` it _must_ parse something or fail, it can't successfully parse nothing. For example, using something like `many1(satisfy(Char.IsWhiteSpace))` for `spacesP`, would only succeed when it parses one or more space.
> 

thank you for your help, I really appreciate.
yes it is just a wrapper call to "spaces" that select a common type for using it with "either" (or "choice" as suggested):

```
var spacesP =
  from _s in spaces
  select new EnumRecord { IsValid = false };
```

Then I am confused, how should I skip lines 1 to 12 in my example, knowing that I should keep comments in lines like this, I mean I can't just ignore all comments, only line or multi lines comments:

```
static INTERFACE_t p_list [] =
{
    {ID_E_P_P_M  , (void*)&e_p_p_m,   sizeof(e_p_p_m)   , ID_FLOAT32, PROD}, //TEXT=""E: IPM"" DU=""ppm""
};
```

**Reply by louthy** [reply]
As I mentioned earlier, you should build a general case token parser.  Something that takes a `Parser<A> p`, parses the `p`, and then strips the spaces _after_.  This will allow you to automatically remove the gaps between tokens.

```c#
    Parser<A> token<A>(Parser<A> p) =>
        from r in p
        from _ in spaces
        select r;
```
You need to make sure any parsers you build, that _could_ have spaces after, are wrapped in `token(p)`.  If you use the `makeTokenParser` function, then you need to make sure you build a token parser like so:

```c#
    Parser<A> token<A>(Parser<A> p) =>
        from r in p
        from _ in lexer.WhiteSpace
        select r;
```
Stop manually dealing with spaces, it's a fools errand, treat your source-code as a sequence of tokens and you'll find it much, much easier.
